File: DD_c_plusplus/Examples/KnapsackInstance/KnapsackInstance.cpp
#include "KnapsackInstance.h"


KnapsackInstance::KnapsackInstance(const string& file_path) {
    read_file(file_path);
}

KnapsackInstance::KnapsackInstance(int* initial_state,
                                   vector<pair<string, vector<int>>> variables,
                                   vector<int> weights,
                                   vector<double> objective_weights,
                                   int right_side_of_restrictions)
    : initial_state(initial_state),
      right_side_of_restrictions(right_side_of_restrictions),
      weights(std::move(weights)),
      objective_weights(std::move(objective_weights)),
      variables(std::move(variables))
{}

void KnapsackInstance::read_file(const string& file_path) {
    ifstream input(file_path);
    if( !input.is_open() ) {
        cerr << "Error: could not open file " << file_path << endl;
        exit(1);
    }

    initial_state = new int(0);
    int variable_length;

    input >> variable_length;

    input >> right_side_of_restrictions;

    for (int j = 0; j < variable_length; j++) {
        int value;
        input >> value;
        weights.push_back(value);
    }


    for (int i = 0; i < variable_length; i++) {
        double value;
        input >> value;
        objective_weights.push_back(value);
    }

    for (int i = 1; i <= variable_length; ++i) {
        string variable_name = "x_" + to_string(i);
        variables.emplace_back(make_pair(variable_name, vector<int>{0,1}));
    }
}

File: DD_python/Examples/KnapsackInstance/KnapsackInstance.py
class KnapsackStructure:
    def __init__(self, file_path: str) -> None:
        """
        SourceCode to store the structure of the knapsack problem.
        """

        self.initial_state: int = None
        self.variables: list[tuple[str, list[int]]] = None
        self.right_side_of_restrictions: int = None
        self.weights: list[int] = None
        self.objective_weights: list[int] = None

        self.read_file(file_path)
    
    def read_file(self, file_path: str) -> None:
        try:
            with open(file_path, 'r') as input_file:
                variable_length: int = int(input_file.readline().strip())
                self.right_side_of_restrictions = int(input_file.readline().strip())

                self.weights = list(map(int, input_file.readline().strip().split()))

                self.objective_weights = list(map(int, input_file.readline().strip().split()))

                self.variables = [("x_" + str(i + 1), [0, 1]) for i in range(variable_length)]
                self.initial_state = 0

        except IOError:
            print(f"Error: could not open file {file_path}")
            exit(1)
    
    def __str__(self):
        return f"KnapsackStructure; Initial State: {self.initial_state}\nVariables: {self.variables}\nRight Side of Restrictions: {self.right_side_of_restrictions}\nWeights: {self.weights}\nObjective Weights: {self.objective_weights}"